File: solutions/day13.py
with open("./input.txt", "r") as txt:
  mac: str = txt.read().rstrip()
  txt.close()

machines = [
    [
        [
            int(i.replace('=', '+').split('+')[-1]) for i in coords.split(', ')
        ]
    for coords in machine.split('\n')]
for machine in mac.split('\n\n')]

def solveLinearEquationInTwoUnknowns(a: int, b: int, c: int, d: int, x: int, y: int) -> tuple:
    m = a * x - b * d
    return (
        (c * x - b * y) // m,
        (a * y - c * d) // m,
        (c * x - b * y) % m + (a * y - c * d) % m == 0
    )

# Part 1
resultsPart1 = 0
for i in range(len(machines)):
    x, y, isReachable = solveLinearEquationInTwoUnknowns(machines[i][0][0], machines[i][1][0], machines[i][2][0], machines[i][0][1], machines[i][1][1], machines[i][2][1])
    resultsPart1 += x * 3 + y if isReachable else 0
print("Part 1:", resultsPart1)

# Part 2
resultsPart2 = 0
for i in range(len(machines)):
    machines[i][2][0] += 10000000000000
    machines[i][2][1] += 10000000000000
    x, y, isReachable = solveLinearEquationInTwoUnknowns(machines[i][0][0], machines[i][1][0], machines[i][2][0], machines[i][0][1], machines[i][1][1], machines[i][2][1])
    resultsPart2 += x * 3 + y if isReachable else 0
print("Part 2:", resultsPart2)

File: solutions/day13.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const machines = raw.split("\n\n").map((machine) =>
  machine.split("\n").map((coords) =>
    coords
      .split(", ")
      .map((item) => item.replace("=", "+").split("+")[1])
      .map(Number)
  )
);
const linearEquations = machines.map((machine) =>
  machine[0].map((_, colIndex) => machine.map((row) => row[colIndex]))
);

function solveLinearEquationInTwoUnknowns(equations: number[][]): {
  x: number;
  y: number;
  isReachable: boolean;
} {
  const [equation1, equation2] = equations;
  const m = equation1[0] * equation2[1] - equation1[1] * equation2[0];
  const solutions = {
    x: (equation1[2] * equation2[1] - equation1[1] * equation2[2]) / m,
    y: (equation1[0] * equation2[2] - equation1[2] * equation2[0]) / m,
  };
  return {
    ...solutions,
    isReachable: Number.isInteger(solutions.x) && Number.isInteger(solutions.y),
  };
}

// Part 1
const resultsPart1 = linearEquations.reduce((acc, equations) => {
  const result = solveLinearEquationInTwoUnknowns(equations);
  return result.isReachable ? acc + result.x * 3 + result.y : acc;
}, 0);
console.log("Part 1:", resultsPart1);

// Part 2
const resultsPart2 = linearEquations.reduce((acc, equations) => {
  const newEquations = equations.map((equation) => [
    ...equation.slice(0, 2),
    equation[2] + 10000000000000,
  ]);
  const result = solveLinearEquationInTwoUnknowns(newEquations);
  return result.isReachable ? acc + result.x * 3 + result.y : acc;
}, 0);
console.log("Part 2:", resultsPart2);
